### PR TITLE
feat: validate user email during import [DHIS2-10407] 

### PR DESCRIPTION
**Pull request by jbee**
### Summary

The immediate problem this PR wants to solve is to add validation to the user `email` property.

This is done by annotating the `User#getEmail()` with:
```java
@Property( PropertyType.EMAIL )
```
This has the desired effect for imports done via the bundle import framework.

Most REST operations, however, do not use the import framework so no validation did occur.
Therefore explicit validation based on the `SchemaValidator` was added to the `AbstractCrudController` as well.

### Property Introspection Refactoring
To test validation based on annotations the property introspection is required. The existing code wasn't very testable. Therefore I extracted a new abstraction `PropertyIntrospector` for which there are a handful of implementations which have been extracted from existing classes:
* `HibernatePropertyIntrospector` for Hibernate persistence configuration related analysis
* `JacksonPropertyIntrospector` for Jackson annotations based analysis
* `PropertyPropertyIntrospector` to handle `@Property`, `@PropertyRange` and `@PropertyTransformer`
* `TranslatablePropertyIntrospector` to handle `@Translatable`
* `NodePropertyIntrospector` for `@NodeAnnotation` based analysis

These can be composed to a chain using `then` utility method like so:
```java
    public DefaultPropertyIntrospectorService( SessionFactory sessionFactory )
    {
        this( new HibernatePropertyIntrospector( sessionFactory )
            .then( new JacksonPropertyIntrospector() )
            .then( new TranslatablePropertyIntrospector() )
            .then( new PropertyPropertyIntrospector() ) );
    }
```
This allows to only use `JacksonPropertyIntrospector` in test scenarios, directly or wrapped as `PropertyIntrospectorService` which can be created by:
```java
new DefaultPropertyIntrospectorService(new JacksonPropertyIntrospector());
```
The only role of the `PropertyIntrospectorService` now is to act as a cache for the used `PropertyIntrospector` within.

### CollectionService Refactoring
When testing the REST API of `AbstractCrudController` to apply validation correctly I noticed that replacement based on `CollectionService#clearCollectionItems` had a tendency to cause exceptions related to persistence session/transactions.
All users of `clearCollectionItems` would do:
```java
clearCollectionItems(...);
addCollectionItems(...);
```
to effectively do a _replace items_. Apart from session issues this also has the problem of opening two separate transactions where the first (clearing) could succeed while the second (adding) could fail leaving the result of a empty collection.
Therefore I decided to change `clearCollectionItems` to `replaceCollectionItems` so that both occur within the same transaction.

### Automatic Testing
A integration test was added that uses the import framework that attempts to update a user's email to an invalid email and verifies that the expected validation error is generated.

A new integration test `DefaultSchemaValidatorTest` was added that makes sure the `DefaultSchemaValidator` that has been cleaned up aggressively continues to behave correct. 

### Manual Testing
The REST API of the `AbstractCrudController` was tested manually. Apart from e2e tests we still lack a base setup to write integration tests for controllers. 

I tested both validation and updating entities by the various means of manipulation. Best was to test this thoroughly is to have a look at the `AbstractCrudController` and go through method by method.

### CODE REVIEW

**Review by mortenoh — APPROVED** [review]

File: dhis-2/dhis-services/dhis-service-schema/src/main/java/org/hisp/dhis/schema/introspection/PropertyPropertyIntrospector.java [hunk 1]
```diff
@@ -0,0 +1,262 @@
...
+import org.hisp.dhis.schema.annotation.Property.Access;
+import org.hisp.dhis.schema.annotation.Property.Value;
+import org.hisp.dhis.schema.annotation.PropertyRange;
+import org.hisp.dhis.schema.annotation.PropertyTransformer;
+import org.springframework.util.Assert;
+
+import com.google.common.primitives.Primitives;
+
+/**
+ * A {@link PropertyIntrospector} adds information to existing {@link Property}
```
Comment (enricocolasante) [inline dhis-2/dhis-services/dhis-service-schema/src/main/java/org/hisp/dhis/schema/introspection/PropertyPropertyIntrospector.java:56]: Should it be PropertyPropertyInspector?

File: dhis-2/dhis-api/src/main/java/org/hisp/dhis/schema/Property.java [hunk 2]
```diff
@@ -837,6 +837,18 @@ public boolean is( PropertyType propertyType )
         return propertyType != null && propertyType.equals( this.propertyType );
     }
 
+    public boolean is( PropertyType... anyOf )
```
Comment (enricocolasante) [inline dhis-2/dhis-api/src/main/java/org/hisp/dhis/schema/Property.java:840]: isAnyOf could be a better name to understand exactly how this method works

**Review by enricocolasante — APPROVED** [review]
Just 2 small suggestions

File: dhis-2/dhis-services/dhis-service-schema/src/main/java/org/hisp/dhis/schema/introspection/PropertyPropertyIntrospector.java (hunk 1 above)
Comment (jbee) [reply dhis-2/dhis-services/dhis-service-schema/src/main/java/org/hisp/dhis/schema/introspection/PropertyPropertyIntrospector.java:56]: Yes it is a bit unfortunate that the interface is `PropertyIntrospector` and this implementation is for the `@Property` family of annotations. So it was intentional to have _property_ twice in the name. Any other ideas are very welcome.

Comment (enricocolasante) [reply dhis-2/dhis-services/dhis-service-schema/src/main/java/org/hisp/dhis/schema/introspection/PropertyPropertyIntrospector.java:56]: I got that, I tried to think to a better name without success.
I was just pointing at the name in the Javadoc

```suggestion
 * A {@link PropertyPropertyIntrospector} adds information to existing {@link Property}
```

Comment (enricocolasante) [reply dhis-2/dhis-services/dhis-service-schema/src/main/java/org/hisp/dhis/schema/introspection/PropertyPropertyIntrospector.java:56]: Never mind, I got what you were saying in the Javadoc.

Comment (jbee) [reply dhis-2/dhis-services/dhis-service-schema/src/main/java/org/hisp/dhis/schema/introspection/PropertyPropertyIntrospector.java:56]: Oh, yes the wording is off, I meant to have a _that_ after the class name there.

File: dhis-2/dhis-api/src/main/java/org/hisp/dhis/schema/Property.java (hunk 2 above)
Comment (jbee) [reply dhis-2/dhis-api/src/main/java/org/hisp/dhis/schema/Property.java:840]: I had addressed this but Lars was faster than me with merging.

**Pull request merged**